File: day_5/puzzle.c
#include<stdio.h>
#include<stdlib.h>
#include<stdbool.h>
#include<stdint.h>
#include<string.h>
#include<limits.h>

#define REALLOC_SIZE 32

// Remove a character from a string
void remove_char(char **input, char c) {
    char *read = *input;
    char *write = *input;

    // Advance read pointer till the end.
    // Only advance write pointer if the char != c
    while (*read) {
        *write = *read++;
        write += (*write!= c);
    }
    *write = '\0';

}

// Remove element at ith index.
// Using this since memmove fails on overlapping memory for me.
void remove_ith_elem(char **input, size_t i) {
    size_t len = strlen(*input);

    for (; i < len - 1; i++) {
        (*input)[i] = (*input)[i+1];
    }
    (*input)[i] = '\0';
}

// Remove opposite poles from the string.
void remove_poles(char **input, size_t *input_len) {
    bool was_reduced = true;
    while (was_reduced == true) {
        was_reduced = false;
        // Go through the string from the end - compare and remove opposite poles.
        // Keep repeating until one pass completes without reduction.
        for(int i = *input_len-1; i > 0; i--) {
            if (i == (int)*input_len-1) {
                continue;
            } else {
                // Check if the letters are a small/caps pair.
                if((*input)[i] - (*input)[i+1] == 32 || (*input)[i] - (*input)[i+1] == -32) {
                    // For some reason memmove doesn't work here.
                    // Could be due to overlapping memory.
                    // Use own function to remove from the string.
                    remove_ith_elem(input, i+1);
                    remove_ith_elem(input, i);
                    *input_len -= 2;
                    was_reduced = true;
                }
            }
        }
    }
}

int main() {

    FILE *fp;
    size_t realloc_counter = REALLOC_SIZE;

    size_t input_len = 0;
    size_t golden_len = 0;
    char *input = malloc(realloc_counter);
    char *golden_input;
    char letter;

    int small_letter = 65;
    int caps_letter = 97;
    int smallest_length;

    // Read the file.
    fp = fopen("./input.txt", "r");
    if (fp == NULL) {
        perror("input.txt");
        exit(EXIT_FAILURE);
    }

    // Store all values in a buffer.
    while ((letter = fgetc(fp)) != EOF) {
        input[input_len++] = (char)letter;
        if(input_len >= realloc_counter) {
            realloc_counter += REALLOC_SIZE;
            input = realloc(input, realloc_counter);
        }
    }
    input[input_len] = '\0';
    fclose(fp);

    // Before making any changes to the input, save a golden copy for part 2.
    golden_len = input_len;
    golden_input = malloc(golden_len);
    strcpy(golden_input, input);

    // Now go ahead and reduce the input to find the final length.
    remove_poles(&input, &input_len);

    // Final length is 1 less than the string length to account for \0
    printf ("Final reduced length = %d\n", (int)input_len - 1);

    // Now start off removing one letter from the string at a time.
    // Then reduce it and find the smallest length.
    smallest_length = strlen(golden_input);
    for (int i = 0; i < 26; i++) {
        // Reset the input on each pass.
        strcpy(input, golden_input);
        // Use ASCII code of letters to remove them from string.
        remove_char(&input, (i+small_letter));
        remove_char(&input, (i+caps_letter));
        input_len = strlen(input);
        remove_poles(&input, &input_len);
        if ((int)input_len < smallest_length) {
            smallest_length = (int)input_len;
        }
    }

    printf ("Final smallest length = %d\n", (int)smallest_length - 1);

    // Free all the allocated memory.
    free(input);
    free(golden_input);

    return 0;
}

File: day_5/puzzle.py
"""
AOC 2018, Day 5
"""

__author__ = "adibis"
__version__ = "0.1.0"
__license__ = "MIT"

# Remove poles (which are back to back letters of opposite case)
def remove_poles(read_data):
    was_reduced = True
    while was_reduced == True:
        was_reduced = False
        for i in range(len(read_data) - 1, -1, -1):
            if i == (len(read_data) - 1):
                continue
            else:
                # Check if letters differ in ASCII code by 32. If yes it's a matching pair.
                # Keep repeating until we have at least one reduction possible.
                if abs(ord(read_data[i]) - ord(read_data[i+1])) == 32:
                    del read_data[i+1]
                    del read_data[i]
                    was_reduced = True

    return read_data

def main():
    """Day five of AOC 2018"""
    with open ("input.txt", 'r') as f:
        read_data = f.read()


    # Save the read data for reuse later.
    golden_read_data = read_data

    # Part 1: Reduce the list as much as possible. Find the final size of it.
    read_data = list(read_data.strip())
    read_data = remove_poles(read_data)

    print ("PART 1: Length of Final String = ", len(read_data))

    # Part 2: Again, use the ascii code of letters to delete them from the string in a loop.
    small_letter = 65
    caps_letter = 97
    smallest_length = len(read_data)

    for alpha in range(0, 25):
        # Reset the input, delete letters one by one.
        # Then repeat the code above to see if it was reduced or not.
        read_data = golden_read_data
        read_data = read_data.replace(chr(small_letter+alpha), "")
        read_data = read_data.replace(chr(caps_letter+alpha), "")
        read_data = list(read_data.strip())
        read_data = remove_poles(read_data)

        if smallest_length > len(read_data):
            smallest_length = len(read_data)

    print ("PART 2: Length of Smallest String = ", smallest_length)


if __name__ == "__main__":
    """ This is executed when run from the command line """
    main()
